internal/api/oidc/token_exchange.go

```diff
@@ -0,0 +1,331 @@
+		if len(subjectAudience) > 0 {
+			return subjectAudience, nil
+		}
+		if len(actorAudience) > 0 {
+			return actorAudience, nil
+		}
+	}
+	if slices.Equal(requestedAudience, subjectAudience) || slices.Equal(requestedAudience, actorAudience) {
+		return requestedAudience, nil
+	}
+	allowedAudience := subjectAudience
+	allowedAudience = append(subjectAudience, actorAudience...)
+	for _, a := range requestedAudience {
+		if !slices.Contains(allowedAudience, a) {
+			return nil, oidc.ErrInvalidTarget().WithDescription("audience %q not found in subject or actor token", a)
+		}
+	}
+	return requestedAudience, nil
+}
+
+// createExchangeTokens prepares the final tokens to be returned to the client.
+// The subjectToken is used to set the new token's subject and resource owner.
+// The actorToken is used to set the new token's auth time AMR and actor.
+// Both tokens may point to the same object (subjectToken) in case of a regular Token Exchange.
+// When the subject and actor Tokens point to different objects, the new tokens will be for impersonation / delegation.
+func (s *Server) createExchangeTokens(ctx context.Context, tokenType oidc.TokenType, client *Client, subjectToken, actorToken *exchangeToken, audience, scopes []string) (_ *oidc.TokenExchangeResponse, err error) {
+	var (
+		userInfo   *oidc.UserInfo
+		signingKey op.SigningKey
+	)
+	if slices.Contains(scopes, oidc.ScopeOpenID) || tokenType == oidc.JWTTokenType || tokenType == oidc.IDTokenType {
+		projectID := client.client.ProjectID
+		userInfo, err = s.userInfo(ctx, subjectToken.userID, projectID, scopes, []string{projectID})
+		if err != nil {
+			return nil, err
+		}
+		signingKey, err = s.Provider().Storage().SigningKey(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	resp := &oidc.TokenExchangeResponse{
+		Scopes: scopes,
+	}
+
+	reason := domain.TokenReasonExchange
+	actor := actorToken.actor
+	if subjectToken != actorToken {
+		reason = domain.TokenReasonImpersonation
+		actor = actorToken.nestedActor()
+	}
+
+	switch tokenType {
+	case oidc.AccessTokenType, "":
+		resp.AccessToken, resp.RefreshToken, resp.ExpiresIn, err = s.createExchangeAccessToken(ctx, client, subjectToken.resourceOwner, subjectToken.userID, audience, scopes, actorToken.authMethods, actorToken.authTime, reason, actor)
+		resp.TokenType = oidc.BearerToken
+		resp.IssuedTokenType = oidc.AccessTokenType
+
+	case oidc.JWTTokenType:
+		resp.AccessToken, resp.RefreshToken, resp.ExpiresIn, err = s.createExchangeJWT(ctx, signingKey, client, subjectToken.resourceOwner, subjectToken.userID, audience, scopes, actorToken.authMethods, actorToken.authTime, reason, actor, userInfo.Claims)
+		resp.TokenType = oidc.BearerToken
+		resp.IssuedTokenType = oidc.JWTTokenType
+
+	case oidc.IDTokenType:
+		resp.AccessToken, resp.ExpiresIn, err = s.createExchangeIDToken(ctx, signingKey, client, subjectToken.userID, "", audience, userInfo, actorToken.authMethods, actorToken.authTime, reason, actor)
+		resp.TokenType = "N_A"
```

Comment:
maybe we should put it into a const or at least add a comment that this is based on the RFC
i already stumbled over it